### PR TITLE
feat: add page transitions rfc

### PR DESCRIPTION
**Pull request by fsdiogo**

### CODE REVIEW

File: rfcs/0004-page-transitions.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+import Highway from '@dogstudio/highway/src/highway';
+import { Slide, Fade } from '../shared/components/transitions';
+
+class App extends NextApp {
+    Highway;
+
+    componentDidMount() {
+        this.unregisterGoogleTracking = registerGoogleTracking(this.props.router);
+
+        this.Highway = new Highway.Core({
```
Comment (satazor) [inline rfcs/0004-page-transitions.md]: Do we need to keep a reference to it? Perhaps we need it to destroy in `componentWillUnmount`? If not, I would get rid of it.

File: rfcs/0004-page-transitions.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+## Enabling Highway
+
+In our [`App.js`](https://github.com/moxystudio/next-with-moxy/blob/master/www/app/App.js) file, we'll have to enable Highway:
+
+```js
+import Highway from '@dogstudio/highway/src/highway';
+import { Slide, Fade } from '../shared/components/transitions';
+
+class App extends NextApp {
+    Highway;
```
Comment (satazor) [inline rfcs/0004-page-transitions.md]:
```suggestion
   highway;
```

File: rfcs/0004-page-transitions.md (hunk 1 above)
Comment (satazor) [inline rfcs/0004-page-transitions.md]:
```suggestion
        this.highway = new Highway.Core({
```

File: rfcs/0004-page-transitions.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+render() {
+    const { Component, pageProps, router } = this.props;
+
+    return (
+        <>
+            <Head>
+                {/* ... */}
+            </Head>
+            <KeyboardOnlyOutlines>
+                <div data-router-wrapper>
```
Comment (satazor) [inline rfcs/0004-page-transitions.md:59]: Is there any way for us to diminish the number of wrappers needed?

File: rfcs/0004-page-transitions.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+class App extends NextApp {
+    Highway;
+
+    componentDidMount() {
+        this.unregisterGoogleTracking = registerGoogleTracking(this.props.router);
+
+        this.Highway = new Highway.Core({
+            transitions: {
+                '/contact': Fade,
```
Comment (satazor) [inline rfcs/0004-page-transitions.md:36]: Is it possible to decide which animation to use depending on the current page? e.g.: When I'm on the homepage and click "Contact", I want the fade in animation to play; however, if I'm on the about page and click "Contact", I want the slide animation to play instead.

**Review by satazor — COMMENTED** [review]
Other questions:

- Does highway handle any scrolling? Does it attempt to restore scroll? (e.g.: when navigating back)
- Related to the previous question: how does scroll restoration plays with async data-fetching that can take longer? Like attempting to restore scroll to an area not yet available because data is still loading.

File: rfcs/0004-page-transitions.md (hunk 1 above)
Comment (fsdiogo) [reply rfcs/0004-page-transitions.md]: In this example we don't need to keep the reference. I just kept it there to show that in some cases we may need it, for example with [events](https://highway.js.org/examples/events.html) probably.

File: rfcs/0004-page-transitions.md (hunk 3 above)
Comment (fsdiogo) [reply rfcs/0004-page-transitions.md:59]: Not that I know of.

File: rfcs/0004-page-transitions.md (hunk 4 above)
Comment (fsdiogo) [reply rfcs/0004-page-transitions.md:36]: I haven't tested it myself but I think so, with [contextual transitions](https://highway.js.org/get-started.html#contextual-transitions):

```js
const H = new Highway.Core({
	renderers: {
		name: CustomRenderer
	},
	transitions: {
		name: CustomTransition,
		contextual: {
			foo: ContextualTransition
		}
	}
});
```

**Comment by fsdiogo** [comment]
> Does highway handle any scrolling? Does it attempt to restore scroll? (e.g.: when navigating back)

It seems that it doesn't restore the scroll position: https://github.com/Dogstudio/highway/issues/52.

> Related to the previous question: how does scroll restoration plays with async data-fetching that can take longer? Like attempting to restore scroll to an area not yet available because data is still loading.

N/A.

**Comment by satazor** [comment]
@fsdiogo that’s one major drawback unfortunately. It seems that the alternative package you mentioned does restore the scroll.

Can you list this in the drawbacks section?

**Comment by fsdiogo** [comment]
Superseded by https://github.com/moxystudio/rfcs-oss/pull/6.

**Pull request closed without merging**